### ISSUE TITLE
Tensorflow cudnn integration status

### ISSUE THREAD

**Issue by Mistobaan** [tensorflow/tensorflow#2937]
With this issue I would like to track the efforts in integrating the
cudnn library within tensorflow.

As of June 17th 2016 doing a manual grep on the repository gives these
functions as being mapped from cudnn to the stream executor,

From chapter 4 of the cudnn User Guide version 5.0 (April 2016):

- [x] cudnnGetVersion
- [ ] cudnnGetErrorString
- [x] cudnnCreate
- [x] cudnnDestroy
- [x] cudnnSetStream
- [ ] cudnnGetStream
- [x] cudnnCreateTensorDescriptor
- [ ] cudnnSetTensor4dDescriptor
- [ ] cudnnSetTensor4dDescriptorEx
- [ ] cudnnGetTensor4dDescriptor
- [x] cudnnSetTensorNdDescriptor
- [x] cudnnDestroyTensorDescriptor
- [x] cudnnTransformTensor
- [x] cudnnAddTensor
- [ ] cudnnOpTensor
- [x] cudnnSetTensor
- [ ] cudnnScaleTensor
- [x] cudnnCreateFilterDescriptor
- [ ] cudnnSetFilter4dDescriptor
- [ ] cudnnGetFilter4dDescriptor
- [ ] cudnnSetFilter4dDescriptor_v3 (versioned)
- [ ] cudnnGetFilter4dDescriptor_v3 (versioned)
- [ ] cudnnSetFilter4dDescriptor_v4 (versioned)
- [ ] cudnnGetFilter4dDescriptor_v4 (versioned)
- [x] cudnnSetFilterNdDescriptor
- [ ] cudnnGetFilterNdDescriptor
- [ ] cudnnGetFilterNdDescriptor_v3 (versioned)
- [ ] cudnnGetFilterNdDescriptor_v3 (versioned)
- [ ] cudnnGetFilterNdDescriptor_v4 (versioned)
- [ ] cudnnGetFilterNdDescriptor_v4 (versioned)
- [x] cudnnDestroyFilterDescriptor
- [x] cudnnCreateConvolutionDescriptor
- [ ] cudnnSetConvolution2dDescriptor
- [ ] cudnnGetConvolution2dDescriptor
- [ ] cudnnGetConvolution2dForwardOutputDim
- [x] cudnnSetConvolutionNdDescriptor
- [ ] cudnnGetConvolutionNdDescriptor
- [x] cudnnGetConvolutionNdForwardOutputDim
- [x] cudnnDestroyConvolutionDescriptor
- [ ] cudnnFindConvolutionForwardAlgorithm
- [ ] cudnnFindConvolutionForwardAlgorithmEx
- [x] cudnnGetConvolutionForwardAlgorithm
- [x] cudnnGetConvolutionForwardWorkspaceSize
- [x] cudnnConvolutionForward
- [ ] cudnnConvolutionBackwardBias
- [ ] cudnnFindConvolutionBackwardFilterAlgorithm
- [ ] cudnnFindConvolutionBackwardFilterAlgorithmEx
- [x] cudnnGetConvolutionBackwardFilterAlgorithm
- [x] cudnnGetConvolutionBackwardFilterWorkspaceSize
- [x] cudnnConvolutionBackwardFilter
- [ ] cudnnFindConvolutionBackwardDataAlgorithm
- [ ] cudnnFindConvolutionBackwardDataAlgorithmEx
- [x] cudnnGetConvolutionBackwardDataAlgorithm
- [x] cudnnGetConvolutionBackwardDataWorkspaceSize
- [x] cudnnConvolutionBackwardData
- [ ] cudnnSoftmaxForward
- [ ] cudnnSoftmaxBackward
- [x] cudnnCreatePoolingDescriptor
- [ ] cudnnSetPooling2dDescriptor
- [ ] cudnnGetPooling2dDescriptor
- [x] cudnnSetPoolingNdDescriptor
- [ ] cudnnGetPoolingNdDescriptor
- [ ] cudnnSetPooling2dDescriptor_v3 (versioned)
- [ ] cudnnGetPooling2dDescriptor_v3 (versioned)
- [ ] cudnnSetPoolingNdDescriptor_v3 (versioned)
- [ ] cudnnGetPoolingNdDescriptor_v3 (versioned)
- [ ] cudnnSetPooling2dDescriptor_v4 (versioned)
- [ ] cudnnGetPooling2dDescriptor_v4 (versioned)
- [ ] cudnnSetPoolingNdDescriptor_v4 (versioned)
- [ ] cudnnGetPoolingNdDescriptor_v4 (versioned)
- [x] cudnnDestroyPoolingDescriptor
- [ ] cudnnGetPooling2dForwardOutputDim
- [ ] cudnnGetPoolingNdForwardOutputDim
- [x] cudnnPoolingForward
- [x] cudnnPoolingBackward
- [x] cudnnActivationForward
- [ ] cudnnActivationBackward
- [x] cudnnCreateActivationDescriptor
- [x] cudnnSetActivationDescriptor
- [x] cudnnGetActivationDescriptor
- [x] cudnnDestroyActivationDescriptor
- [ ] cudnnActivationForward_v3  (versioned)
- [ ] cudnnActivationBackward_v3 (versioned)
- [ ] cudnnActivationForward_v4  (versioned)
- [ ] cudnnActivationBackward_v4 (versioned)
- [ ] cudnnCreateLRNDescriptor
- [ ] cudnnSetLRNDescriptor
- [ ] cudnnGetLRNDescriptor
- [ ] cudnnDestroyLRNDescriptor
- [ ] cudnnLRNCrossChannelForward
- [ ] cudnnLRNCrossChannelBackward
- [ ] cudnnDivisiveNormalizationForward
- [ ] cudnnDivisiveNormalizationBackward
- [ ] cudnnBatchNormalizationForwardInference
- [ ] cudnnBatchNormalizationForwardTraining
- [ ] cudnnBatchNormalizationBackward
- [ ] cudnnDeriveBNTensorDescriptor
- [ ] cudnnCreateRNNDescriptor
- [ ] cudnnDestroyRNNDescriptor
- [ ] cudnnSetRNNDescriptor
- [ ] cudnnGetRNNWorkspaceSize
- [ ] cudnnGetRNNTrainingReserveSize
- [ ] cudnnGetRNNParamsSize
- [ ] cudnnGetRNNLinLayerMatrixParams
- [ ] cudnnGetRNNLinLayerBiasParams
- [ ] cudnnRNNForwardInference
- [ ] cudnnRNNForwardTraining
- [ ] cudnnRNNBackwardData
- [ ] cudnnRNNBackwardWeights
- [ ] cudnnCreateDropoutDescriptor
- [ ] cudnnDestroyDropoutDescriptor
- [ ] cudnnDropoutGetStatesSize
- [ ] cudnnDropoutGetReserveSpaceSize
- [ ] cudnnSetDropoutDescriptor
- [ ] cudnnDropoutForward
- [ ] cudnnDropoutBackward
- [ ] cudnnCreateSpatialTransformerDescriptor
- [ ] cudnnDestroySpatialTransformerDescriptor
- [ ] cudnnSetSpatialTransformerNdDescriptor
- [ ] cudnnSpatialTfGridGeneratorForward
- [ ] cudnnSpatialTfGridGeneratorBackward
- [ ] cudnnSpatialTfSamplerForward
- [ ] cudnnSpatialTfSamplerBackward

### Batch Normalization
Seems @lukemetz was working on it but has stalled for a bit https://github.com/tensorflow/tensorflow/pull/1759

### What is the plan for the RNN ?
I know @wchan was working on a [cpu version](https://github.com/tensorflow/tensorflow/pull/2002). I was trying to get a stub at [using the cudnn version](https://github.com/Mistobaan/tensorflow/tree/feature/cudnn-rnn-lstm) but from the comments in that thread seemed like @zheng-xq was working on it internally.
Can anyone comment on the status of these ops?

### Other Questions

1. Any reasons why the Softmax functions are not being used ?
1. Would make sense to split the above list in chunks and create issues with contribution welcome, so external contributors can tackle them without duplicating internal work, or mark what google will be working on internally ?

I hope this helps in organizing the work around the cudnn and inspire the community to contribute. I will try to keep this issue up to date.

**Comment by zheng-xq** [comment]
In general, TensorFlow prefers its own implementation on GPU, if the performance is close enough. This is because we would have the source code for customization. Conv is a good example where it would take a lot of effort to replicate Cudnn's performance. 

There is still an ongoing investigation how TensorFlow would expose a fused RNN/LSTM implementations. The current Cudnn interface does not provide enough customization to accommodate common variations. In the mean time, it is possible that we can add a wrapper op in TensorFlow/contrib around the current Cudnn implementation, before we can decide what is the best way for long-term support in the TensorFlow core.

**Comment by bhack** [comment]
@zheng-xq How much of this policy is related to the blend effort of each framework to an API conformation on cudnn and how much to the control  of  kernels code from open source point of view?

**Comment by zheng-xq** [comment]
@bhack, in case of LSTM, the biggest concern is lack of customization. In general, having access to the source code gives us better ability us to tweak our code to better serve TensorFlow users. 

That said, we won't hesitate to pick any good functionalities from any library including Cudnn, if there is evidence showing a sizeable benefit.

**Comment by michaelisard** [comment]
Closing this tracking bug because it seems there is not a goal to map all cuDNN functions for reasons given by @zheng-xq. Please open separate issues to track specific missing cuDNN functionality.

**Comment by bhack** [comment]
@michaelisard Can you give me an estimated ratio between not satisfied API design and the closed source nature of cudnn that not attracted full coverage?

**Comment by michaelisard** [comment]
@zheng-xq can comment better than I can.

**Comment by bhack** [comment]
You can close. I was only curious.

**Comment by Mistobaan** [comment]
I would be interested in having full support of the cuDNN library through the stream executor and all the goodies that comes with it. Expose all the endpoints in the cudnn layer and only few to the public layer. 

@zheng-xq  did the problem with cudnnRNN got solved at the end? When I tried the wrapping I wasn't sure where to store the temporary variables required by the cudnn library.

**Comment by zheng-xq** [comment]
@bhack, I don't think anyone sets a clear ratio. And it varies case by case for each op. For new ops, we firstly prefer a fast solution with source within reasonable amount of effort. If it takes a lot of effort to match the performance of Cudnn, or any other library, we would not hesitate to use that library. Overall, we go with the fastest solution. If the performance is about the same between our custom code and Cudnn, we prefer the one with the source.

@Mistobaan, we are working on adding a Cudnn RNN layer to tf.contrib. We will announce in the release note when it is done.
